internal/releasesources/release_sources.go

```
@@ -0,0 +1,510 @@
+	mergedReleaseImages := []*common.ReleaseImage{}
+	staticReleaseImagesReferenceSet := map[string]bool{}
+	defaultStaticReleaseExists := false
+
+	for _, staticRelease := range staticReleaseImages {
+		mergedReleaseImages = append(mergedReleaseImages, staticRelease)
+		staticReleaseImagesReferenceSet[*staticRelease.URL] = true
+		if staticRelease.Default {
+			defaultStaticReleaseExists = true
+		}
+	}
+
+	for _, dynamicRelease := range dynamicReleaseImages {
+		if staticReleaseImagesReferenceSet[*dynamicRelease.URL] {
+			continue
+		}
+		if dynamicRelease.Default && defaultStaticReleaseExists {
+			dynamicRelease.Default = false
+		}
+		mergedReleaseImages = append(mergedReleaseImages, dynamicRelease)
+	}
+
+	return mergedReleaseImages
+}
+
+func (h *releaseSourcesHandler) syncReleaseImagesWithErr() error {
+	err := h.validateReleaseSources()
+	if err != nil {
+		return err
+	}
+
+	err = h.validateStaticReleaseImages()
+	if err != nil {
+		return err
+	}
+	dynamicDBReleaseImages, err := h.getDynamicReleaseImages()
+	if err != nil {
+		return err
+	}
+	h.log.Debugf("Found %d dynamic release Images", len(dynamicDBReleaseImages))
+	supportLevels, err := h.getSupportLevels()
+	if err != nil {
+		return err
+	}
+	err = h.setSupportLevels(dynamicDBReleaseImages, supportLevels)
+	if err != nil {
+		return err
+	}
+	staticDBReleaseImages := h.getStaticReleaseImages()
+	h.log.Debugf("Found %d static release Images", len(staticDBReleaseImages))
+	err = h.setSupportLevels(staticDBReleaseImages, supportLevels)
+	if err != nil {
+		return err
+	}
+	err = h.setDefaultReleaseImage(dynamicDBReleaseImages)
+	if err != nil {
+		return err
+	}
+
+	dynamicDBReleaseImages, err = h.removeReleaseImageDuplicates(dynamicDBReleaseImages)
+	if err != nil {
+		return err
+	}
+
+	dbReleaseImages := h.mergeDBReleaseImages(staticDBReleaseImages, dynamicDBReleaseImages)
+
+	h.log.Debug("Starting SQL transaction")
+	tx := h.db.Begin()
```

why not to use same template we use at all other places 
`txSuccess := false
	tx := h.db.Begin()
	defer func() {
		if !txSuccess {
			log.Error("update monitored operator failed")
			tx.Rollback()
		}
		if r := recover(); r != nil {
			log.Error("update monitored operator failed")
			tx.Rollback()
		}
	}()`
no idea though why we are not making it a function to be reusable

---

I think the main issue with this is the `recover`, I had a talk about it with @filanov and @rccrdpccl, we came to a conclusion that hides errors, therefore we should not recover.

---

We can create a function without recovering though

---

all the recovers should already be removed, we replaced the code with using a new format that will fail the transaction and handle the rollback https://github.com/openshift/assisted-service/blob/a8dca1f6d37e0e17b37c93e1c24426fc9f80ff1c/internal/bminventory/inventory.go#L1288 
otherwise this recover will return a `nil, nil` as a replay and the caller will think that everything is ok